hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java

```diff
@@ -243,34 +250,57 @@ public static HoodieMergedLogRecordScanner.Builder newBuilder() {
   @Override
   public <T> void processNextRecord(HoodieRecord<T> newRecord) throws IOException {
     String key = newRecord.getRecordKey();
-    HoodieRecord<T> prevRecord = records.get(key);
+    HoodieMergeKey mergeKey = newRecord.getMergeKey();
+    if (mergeKey == null) {
+      // If mergeKey is null, then create a simple merge key using record key
+      mergeKey = new HoodieSimpleMergeKey(newRecord.getKey());
+    }
+    HoodieRecord<T> prevRecord = records.get(mergeKey);
     if (prevRecord != null) {
       // Merge and store the combined record
-      HoodieRecord<T> combinedRecord = (HoodieRecord<T>) recordMerger.merge(prevRecord, readerSchema,
-          newRecord, readerSchema, this.getPayloadProps()).get().getLeft();
-      // If pre-combine returns existing record, no need to update it
-      if (combinedRecord.getData() != prevRecord.getData()) {
-        HoodieRecord latestHoodieRecord =
-            combinedRecord.newInstance(new HoodieKey(key, newRecord.getPartitionPath()), newRecord.getOperation());
-
-        latestHoodieRecord.unseal();
-        latestHoodieRecord.setCurrentLocation(newRecord.getCurrentLocation());
-        latestHoodieRecord.seal();
-
-        // NOTE: Record have to be cloned here to make sure if it holds low-level engine-specific
-        //       payload pointing into a shared, mutable (underlying) buffer we get a clean copy of
-        //       it since these records will be put into records(Map).
-        records.put(key, latestHoodieRecord.copy());
+      try {
+        HoodieMergeKey finalMergeKey = mergeKey;
+        recordMerger.fullOuterMerge(prevRecord, readerSchema, newRecord, readerSchema, this.getPayloadProps()).forEach(
+            mergedRecord -> {
+              HoodieRecord<T> combinedRecord = mergedRecord.getLeft();
+              if (combinedRecord.getData() != prevRecord.getData()) {
+                HoodieRecord latestHoodieRecord = getLatestHoodieRecord(newRecord, combinedRecord, key);
+                records.put(finalMergeKey, latestHoodieRecord.copy());
+              }
+            });
+      } catch (UnsupportedOperationException e) {
+        LOG.warn("Error merging records: Full outer merge not supported, falling back to default merge for key {}", key);
+        HoodieRecord<T> combinedRecord = (HoodieRecord<T>) recordMerger.merge(prevRecord, readerSchema,
+            newRecord, readerSchema, this.getPayloadProps()).get().getLeft();
+        // If pre-combine returns existing record, no need to update it
+        if (combinedRecord.getData() != prevRecord.getData()) {
```

Comment:
I don't think we should put any regular code path into a catch block.

Reply:
Basically, want to fallback to original method if a merger is used that does not support MergeKey based merging. Wanted to avoid another boolean flag in the scanner so as not to expose the implementation detail here. Let me think if there can be a better way to do this.

Reply:
This code is not obsolete. Have changed as per our offline dicussion.